# AAC Siegfried Analysis

This notebook explores some of the research questions arising following the analysis of the British Library Asian and African Studies collection metadata with [Siegfried](https://www.itforarchivists.com/siegfried/)

During August 2019, a total of 69.33 TB of material was analysed. The combined DROID-style output for TIFF files from Siegfried resulting from this analysis is available in `CombinedOutput_2019-09-03_allTiffs_parentDirs_folderCount.csv` - the original output from Siegfried has been augmented to include only TIFF files and the parent directory of each file scanned. The `FolderCount` field was added in order to find the number of files present in each folder.

An important note: the following pandas dataframes only exist within memory and do not persist without writing them to a file!

## Setting up an analysis environment
***This section describes setting up a python environemt with the pip package manager - others are available***
### You will need: 
- Python 3 (already installed if you can see this!)
- Pip package manager.

At the very minimum you will probably need to have installed the pandas and numpy python modules. Others, such as matplotlib, will be necessary for any data visualisation.

Install pip with:

In [ ]:
#curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
#python get-pip.py

Install packages with

In [ ]:
#pip install numpy pandas #and any others...

Import the necessary python modules:

In [1]:
import pandas as pd 
import numpy as np
pd.set_option('display.max_columns', None)

Load the csv file as a pandas DataFrame and print a few rows:

In [2]:
allData=pd.read_csv("CombinedOutput_2019-09-03_allTiffs_parentDirs_folderCount.csv",sep=',')
allData.head(5)

ID  PARENT_ID                                                URI  \
0   1        NaN  file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1   2        NaN  file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2   3        NaN  file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3   4        NaN  file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4   5        NaN  file:///P12L-NAS5/Persian%20Digitisation/Grenv...   

                                           FILE_PATH                     NAME  \
0  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001r.tif   
1  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001v.tif   
2  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002r.tif   
3  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002v.tif   
4  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f003r.tif   

      METHOD STATUS        SIZE  TYPE  EXT              LAST_MODIFIED  \
0  Signature   Done  90714782.0  File  tif  2013-04-22T10:40:14+01:00   
1  Signature   Done  85295734.0  File  tif  2013-04-25T10:57:52+01:00   
2  Signature   Done  89390472.0  File  tif  2013-04-22T10:40:32+01:00   
3  Signature   Done  82452820.0  File  tif  2013-04-22T11:11:26+01:00   
4  Signature   Done  86180984.0  File  tif  2013-04-22T10:40:44+01:00   

   EXTENSION_MISMATCH                                        SHA256_HASH  \
0               False  8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...   
1               False  256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...   
2               False  9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...   
3               False  a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...   
4               False  9429574137662b13efa5256290a75982f9c3ec60245233...   

   FORMAT_COUNT       PUID   MIME_TYPE  \
0           1.0  x-fmt/387  image/tiff   
1           1.0  x-fmt/387  image/tiff   
2           1.0  x-fmt/387  image/tiff   
3           1.0  x-fmt/387  image/tiff   
4           1.0  x-fmt/387  image/tiff   

                                     FORMAT_NAME  FORMAT_VERSION  \
0  Exchangeable Image File Format (Uncompressed)             2.2   
1  Exchangeable Image File Format (Uncompressed)             2.2   
2  Exchangeable Image File Format (Uncompressed)             2.2   
3  Exchangeable Image File Format (Uncompressed)             2.2   
4  Exchangeable Image File Format (Uncompressed)             2.2   

                                          PARENT_DIR  FolderCount  
0  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
1  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
2  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
3  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
4  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1

Remove `NaN` (Not A Number) null values and replace them with an empty string "" to enable some string comparisons later down the line

In [3]:
allData=allData.replace(np.nan,"")
allData.head(5)

ID PARENT_ID                                                URI  \
0   1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1   2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2   3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3   4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4   5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   

                                           FILE_PATH                     NAME  \
0  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001r.tif   
1  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001v.tif   
2  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002r.tif   
3  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002v.tif   
4  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f003r.tif   

      METHOD STATUS        SIZE  TYPE  EXT              LAST_MODIFIED  \
0  Signature   Done  90714782.0  File  tif  2013-04-22T10:40:14+01:00   
1  Signature   Done  85295734.0  File  tif  2013-04-25T10:57:52+01:00   
2  Signature   Done  89390472.0  File  tif  2013-04-22T10:40:32+01:00   
3  Signature   Done  82452820.0  File  tif  2013-04-22T11:11:26+01:00   
4  Signature   Done  86180984.0  File  tif  2013-04-22T10:40:44+01:00   

   EXTENSION_MISMATCH                                        SHA256_HASH  \
0               False  8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...   
1               False  256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...   
2               False  9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...   
3               False  a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...   
4               False  9429574137662b13efa5256290a75982f9c3ec60245233...   

   FORMAT_COUNT       PUID   MIME_TYPE  \
0           1.0  x-fmt/387  image/tiff   
1           1.0  x-fmt/387  image/tiff   
2           1.0  x-fmt/387  image/tiff   
3           1.0  x-fmt/387  image/tiff   
4           1.0  x-fmt/387  image/tiff   

                                     FORMAT_NAME FORMAT_VERSION  \
0  Exchangeable Image File Format (Uncompressed)            2.2   
1  Exchangeable Image File Format (Uncompressed)            2.2   
2  Exchangeable Image File Format (Uncompressed)            2.2   
3  Exchangeable Image File Format (Uncompressed)            2.2   
4  Exchangeable Image File Format (Uncompressed)            2.2   

                                          PARENT_DIR  FolderCount  
0  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
1  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
2  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
3  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1  
4  \\P12L-NAS5\Persian Digitisation\Grenville xli...            1

## Duplicate Analysis

Firstly, find exact duplicates by SHA256 hash match. Read this as "show allData where duplicate entries are found in the `SHA256_HASH` column". The [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) shows that the `duplicated` member function defaults to marking duplicates as `True` except for the first occurrence. If you want to include the "original" file in addition to any duplicates, include `keep=False` in the function call. Using the default behaviour, as below, shows only "pure duplicates" and omits the original file.

In [12]:
onlyDuplicates=allData[allData.duplicated(subset="SHA256_HASH")]
onlyDuplicates

ID PARENT_ID                                                URI  \
16637      16638            file:///P12L-NAS5/Persian%20Digitisation/Or%20...   
33491      33499            file:///P12L-NAS5/Preservation%20Digitisation%...   
33492      33500            file:///P12L-NAS5/Preservation%20Digitisation%...   
33493      33501            file:///P12L-NAS5/Preservation%20Digitisation%...   
33494      33502            file:///P12L-NAS5/Preservation%20Digitisation%...   
...          ...       ...                                                ...   
1275916  1328311            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275917  1328312            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275918  1328313            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275919  1328314            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275920  1328315            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
16637    \\P12L-NAS5\Persian Digitisation\Or 5302\or_53...   
33491    \\P12L-NAS5\Preservation Digitisation Bids 201...   
33492    \\P12L-NAS5\Preservation Digitisation Bids 201...   
33493    \\P12L-NAS5\Preservation Digitisation Bids 201...   
33494    \\P12L-NAS5\Preservation Digitisation Bids 201...   
...                                                    ...   
1275916  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275917  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275918  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275919  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275920  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                           NAME     METHOD STATUS         SIZE  TYPE  EXT  \
16637         or_5302_f_1~r.tif  Signature   Done  103677788.0  File  tif   
33491    add_ms_14373_f001r.tif  Signature   Done  101107818.0  File  tif   
33492    add_ms_14373_f001v.tif  Signature   Done   95056216.0  File  tif   
33493    add_ms_14373_f002r.tif  Signature   Done  101107818.0  File  tif   
33494    add_ms_14373_f002v.tif  Signature   Done   95056216.0  File  tif   
...                         ...        ...    ...          ...   ...  ...   
1275916      or_7458_fbrigr.tif  Signature   Done   66538396.0  File  tif   
1275917      or_7458_fbrigv.tif  Signature   Done   66538396.0  File  tif   
1275918       or_7458_fbspi.tif  Signature   Done   39485948.0  File  tif   
1275919     or_7458_fse002r.tif  Signature   Done   66538396.0  File  tif   
1275920     or_7458_fse002v.tif  Signature   Done   66538396.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
16637    2013-06-11T09:44:37+01:00               False   
33491    2017-08-14T11:35:20+01:00               False   
33492    2017-08-14T11:40:32+01:00               False   
33493    2017-08-14T11:35:28+01:00               False   
33494    2017-08-14T11:40:34+01:00               False   
...                            ...                 ...   
1275916       2007-01-04T00:42:21Z               False   
1275917       2007-01-04T00:44:31Z               False   
1275918       2007-01-04T00:46:28Z               False   
1275919       2007-01-04T00:38:13Z               False   
1275920       2007-01-04T00:40:08Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
16637    edc98acdd02aeb5f113c8ccab68e17f4d11a0012d98378...           1.0   
33491    e8d915613378b78a5fa00db97a69acecfc1be50d74bfcb...           1.0   
33492    50ba9310a26f3d72eef400eee3c0464c2b06ac8edbdec0...           1.0   
33493    c0479b980e3d9d68be8f9f246ef51d1d47d822ab774b17...           1.0   
33494    f2de0c0da8f67dbf34b4fb1bc6b2849eaa1f297e60ad87...           1.0   
...                                                    ...           ...   
1275916  0c9f2baf3dd8477bd20327df6f0d3733c8392878512110...           1.0   
1275917  8cc44b99495ca7e7767600d1ba26edb385fa8232826978...        

Results in 85119 pure duplicates found. Finding more information about the duplicates:

In [19]:
totalDuplicateSize=onlyDuplicates['SIZE'].sum()/(1024**4) # Siegfried returns size in bytes, so display size in TB
print("Total size of duplicates: {} TB".format(totalDuplicateSize))

Total size of duplicates: 5.383706840476407 TB


In [28]:
duplicateFileList=onlyDuplicates.index.to_list()

Now we'd like to identify files in the Siegfried output with exactly duplicates. We don't know (yet) which file is the original, so we need to find all files with exact hash matches, rather than setting aside an 'original' set.

In [33]:
duplicatesAndOriginals=allData[allData.duplicated(subset="SHA256_HASH",keep=False)]
duplicatesAndOriginals

ID PARENT_ID                                                URI  \
0              1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1              2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2              3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3              4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4              5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
...          ...       ...                                                ...   
1275916  1328311            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275917  1328312            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275918  1328313            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275919  1328314            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275920  1328315            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
0        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
2        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
3        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
4        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
...                                                    ...   
1275916  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275917  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275918  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275919  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275920  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                            NAME     METHOD STATUS        SIZE  TYPE  EXT  \
0        grenville_xli_f001r.tif  Signature   Done  90714782.0  File  tif   
1        grenville_xli_f001v.tif  Signature   Done  85295734.0  File  tif   
2        grenville_xli_f002r.tif  Signature   Done  89390472.0  File  tif   
3        grenville_xli_f002v.tif  Signature   Done  82452820.0  File  tif   
4        grenville_xli_f003r.tif  Signature   Done  86180984.0  File  tif   
...                          ...        ...    ...         ...   ...  ...   
1275916       or_7458_fbrigr.tif  Signature   Done  66538396.0  File  tif   
1275917       or_7458_fbrigv.tif  Signature   Done  66538396.0  File  tif   
1275918        or_7458_fbspi.tif  Signature   Done  39485948.0  File  tif   
1275919      or_7458_fse002r.tif  Signature   Done  66538396.0  File  tif   
1275920      or_7458_fse002v.tif  Signature   Done  66538396.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
0        2013-04-22T10:40:14+01:00               False   
1        2013-04-25T10:57:52+01:00               False   
2        2013-04-22T10:40:32+01:00               False   
3        2013-04-22T11:11:26+01:00               False   
4        2013-04-22T10:40:44+01:00               False   
...                            ...                 ...   
1275916       2007-01-04T00:42:21Z               False   
1275917       2007-01-04T00:44:31Z               False   
1275918       2007-01-04T00:46:28Z               False   
1275919       2007-01-04T00:38:13Z               False   
1275920       2007-01-04T00:40:08Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
0        8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...           1.0   
1        256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...           1.0   
2        9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...           1.0   
3        a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...           1.0   
4        9429574137662b13efa5256290a75982f9c3ec60245233...           1.0   
...                                                    ...           ...   
1275916  0c9f2baf3dd8477bd20327df6f0d3733c8392878512110...           1.0   
1275917  8cc44b99495ca7e7767600d1ba26edb385fa8232826978...        

Now create a column in the original DataFrame to denote if a file is a duplicate:

In [34]:
allData['IsDuplicate']=False

You can check if this column has been added to the DataFrame with:

In [36]:
allData.columns

Index(['ID', 'PARENT_ID', 'URI', 'FILE_PATH', 'NAME', 'METHOD', 'STATUS',
       'SIZE', 'TYPE', 'EXT', 'LAST_MODIFIED', 'EXTENSION_MISMATCH',
       'SHA256_HASH', 'FORMAT_COUNT', 'PUID', 'MIME_TYPE', 'FORMAT_NAME',
       'FORMAT_VERSION', 'PARENT_DIR', 'FolderCount', 'IsDuplicate'],
      dtype='object')

Turn the ID column from the duplicate dataframe into a python list

In [48]:
duplicateIDList=duplicatesAndOriginals.ID.to_list()

Locate these IDs in the main dataframe and set their `IsDuplicate` value to `True`

In [60]:
allData.loc[allData['ID'].isin(duplicateIDList),'IsDuplicate']=True

Check things have worked:

In [61]:
allData[allData['IsDuplicate']]

ID PARENT_ID                                                URI  \
0              1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1              2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2              3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3              4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4              5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
...          ...       ...                                                ...   
1275916  1328311            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275917  1328312            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275918  1328313            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275919  1328314            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275920  1328315            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
0        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
2        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
3        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
4        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
...                                                    ...   
1275916  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275917  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275918  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275919  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275920  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                            NAME     METHOD STATUS        SIZE  TYPE  EXT  \
0        grenville_xli_f001r.tif  Signature   Done  90714782.0  File  tif   
1        grenville_xli_f001v.tif  Signature   Done  85295734.0  File  tif   
2        grenville_xli_f002r.tif  Signature   Done  89390472.0  File  tif   
3        grenville_xli_f002v.tif  Signature   Done  82452820.0  File  tif   
4        grenville_xli_f003r.tif  Signature   Done  86180984.0  File  tif   
...                          ...        ...    ...         ...   ...  ...   
1275916       or_7458_fbrigr.tif  Signature   Done  66538396.0  File  tif   
1275917       or_7458_fbrigv.tif  Signature   Done  66538396.0  File  tif   
1275918        or_7458_fbspi.tif  Signature   Done  39485948.0  File  tif   
1275919      or_7458_fse002r.tif  Signature   Done  66538396.0  File  tif   
1275920      or_7458_fse002v.tif  Signature   Done  66538396.0  File  tif   

         ... EXTENSION_MISMATCH  \
0        ...              False   
1        ...              False   
2        ...              False   
3        ...              False   
4        ...              False   
...      ...                ...   
1275916  ...              False   
1275917  ...              False   
1275918  ...              False   
1275919  ...              False   
1275920  ...              False   

                                               SHA256_HASH FORMAT_COUNT  \
0        8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...          1.0   
1        256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...          1.0   
2        9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...          1.0   
3        a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...          1.0   
4        9429574137662b13efa5256290a75982f9c3ec60245233...          1.0   
...                                                    ...          ...   
1275916  0c9f2baf3dd8477bd20327df6f0d3733c8392878512110...          1.0   
1275917  8cc44b99495ca7e7767600d1ba26edb385fa8232826978...          1.0   
1275918  7b5a24a82b9d25e08e4884dea94e1b06e9c5ee6cdebfc8...          1.0   
1275919  eea8c2788a5d8e907c3e99bf5a833ac3c6fae4be96780a...          1.0   
1275920  0a3bc9d298bc5f89070117f5f427f44f4ef0c1b779bf0c...          1.0   

              PUID   MIME_TYPE                   

In [62]:
allData.head(10)

ID PARENT_ID                                                URI  \
0   1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1   2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2   3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3   4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4   5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
5   6            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
6   7            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
7   8            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
8   9            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
9  10            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   

                                           FILE_PATH                     NAME  \
0  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001r.tif   
1  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f001v.tif   
2  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002r.tif   
3  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f002v.tif   
4  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f003r.tif   
5  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f003v.tif   
6  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f004r.tif   
7  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f004v.tif   
8  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f005r.tif   
9  \\P12L-NAS5\Persian Digitisation\Grenville xli...  grenville_xli_f005v.tif   

      METHOD STATUS        SIZE  TYPE  EXT  ... EXTENSION_MISMATCH  \
0  Signature   Done  90714782.0  File  tif  ...              False   
1  Signature   Done  85295734.0  File  tif  ...              False   
2  Signature   Done  89390472.0  File  tif  ...              False   
3  Signature   Done  82452820.0  File  tif  ...              False   
4  Signature   Done  86180984.0  File  tif  ...              False   
5  Signature   Done  82452820.0  File  tif  ...              False   
6  Signature   Done  86375264.0  File  tif  ...              False   
7  Signature   Done  82452820.0  File  tif  ...              False   
8  Signature   Done  84313412.0  File  tif  ...              False   
9  Signature   Done  82452820.0  File  tif  ...              False   

                                         SHA256_HASH FORMAT_COUNT       PUID  \
0  8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...          1.0  x-fmt/387   
1  256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...          1.0  x-fmt/387   
2  9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...          1.0  x-fmt/387   
3  a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...          1.0  x-fmt/387   
4  9429574137662b13efa5256290a75982f9c3ec60245233...          1.0  x-fmt/387   
5  d1f825df3279ef4bbaf61fa924013675d41b832381df73...          1.0  x-fmt/387   
6  0ecbf845f0fb4048159a8bd5f398f45dec9c0a408d6652...          1.0  x-fmt/387   
7  57bf87ae9aa9ae78527b9f11e76b96ea7fcb166fa1c27f...          1.0  x-fmt/387   
8  2d2b5cb12ba35953b5bc7c8face86e961e5ae64eb47c4f...          1.0  x-fmt/387   
9  2c21bc9212783746268ed5dd168a44163bdfebd8d47420...          1.0  x-fmt/387   

    MIME_TYPE                                    FORMAT_NAME FORMAT_VERSION  \
0  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
1  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
2  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
3  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
4  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
5  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
6  image/tiff  Exchangeable Image File Format (Uncompressed)            2.2   
7  image/tiff  Exchangeable Image File Format (Uncompres

### Combine IAMS metadata into master spreadsheet

This section uses the optional package [tqdm](https://github.com/tqdm/tqdm), which just adds a nice progress bar/timer to loops for free. 

First, import necessary modules and open the excel worksheets

In [1]:
import numpy as np
import pandas as pd 
from tqdm import tqdm
ms=pd.read_excel("AAC_CombinedAnalysis.xlsx",dtype=object) # set all dtypes to str
iam=pd.read_excel("AAC_IAMSReport.xlsx",dtype=object) 


Replace any null values found with a blank string `""`

In [2]:
ms=ms.replace(np.nan,"")
iam=iam.replace(np.nan,"")

You could now loop over each row of both dataframes and perform string comparisons on the Shelfmark columns. However, even using the `iterrows` member function documented [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html) makes this incredibly slow. I timed this on my laptop to come out somewhere around 40 minutes, and I managed to have it crash near the end twice in a row. 

Pandas DataFrames are essentially 2D arrays anyway and using [numpy arrays](https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html) for this string comparison works out much faster, getting the whole thing done in well under a minute. This does mean we lose the dataframe column headers though, so need to grab those first before converting things to 2D numpy arrays.

In [3]:
msCols=ms.columns.to_list()
iamCols=iam.columns.to_list()
iamColsMod=[]
for col in iamCols:
    iamColsMod.append("IAMS - "+col)

allCols=msCols+iamColsMod

Here I've prepended `IAMS - ` to each column heading from the IAMS report. 

Now convert both dataframes to 2D numpy arrays with dtype object*

**If you convert to dtype=str this places a 10-character limit on each value by default, which is not what you want*

In [4]:
msn=ms.to_numpy(dtype=object)
iamn=iam.to_numpy(dtype=object)

We'd like an overall dataframe that includes values from both of the ones we already have, matched on the shelfmark. So make a blank 2D numpy array with the same number of rows as the master spreadsheet and the same number of columns as both spreadsheets combined. The `empty` function gives you an array of zeros in the shape you specify, all of `dtype=numpy.float64`. Since we've already defined the `msn` and `iamn` arrays as `dtype=object`, we don't want this. Also, to stop just inputting loads of zeros into the eventual spreadsheet, I convert all values to `NaN` before converting the entire array to `dtype=object`.

In [5]:
combined=np.empty([msn.shape[0],(iamn.shape[1]+msn.shape[1])])
combined[:]=np.nan
combined=combined.astype(object)

If there's no match from any row in the IAMS spreadsheet for a particular row in the master spreadsheet, we still want to input some blank cells for the IAMS report columns. So now construct a blank 2D array with the same shape as the IAMS report array, again of dtype=object. 

In [6]:
blanks=np.empty(iamn.shape,dtype=object)

Now loop over both `msn` and `iamn` and find matches on the Shelfmark variable. I checked myself and for any row `n` this corresponds to `msn[n][0]` and `iamn[n][3]`, but this is something you can check yourself!

This is roughly what the next block of code is doing, line-by-line:

- Start with a for loop over `msn`
- initialise `rowMatch` as `False`
- Then start looping over `iamn`
- An `if` statement to compare two strings. 
    - First take the value of shelfmark from `iamn`, remove any whitespace and use `.casefold()` to convert everything to lower-case.
    - Then compare this to the shelfmark from `msn`, with the same whitespace and upper case removal
    - This is just a case-independent string comparison!
- If a Shelfmark match is found:
    - Set `rowMatch=True`
    - Change the row in the large 'combined' 2D array we made to equal all of the values from the current row in `msn` with all of the values from this particular `iamn` row stuck on the end
- If, after checking all of the rows in `iamn`, no match is found:
    - Change the row in the large 'combined' 2D array we made to equal all of the values from the current row in `msn` with blank values added to the end.
- Once all the rows in `msn` have been checked, the `combined` 2D array is no longer blank!

In [7]:
for i in tqdm(range(0,len(msn))):
    rowMatch=False
    for j in range(0,len(iamn)):
        if iamn[j][3].replace(" ","").casefold() == msn[i][0].replace(" ","").casefold():
            rowMatch=True
            combined[i]=np.concatenate([msn[i],iamn[j]])
    if not rowMatch:
        combined[i]=np.concatenate([msn[i],blanks[0]])

100%|██████████| 7634/7634 [00:28<00:00, 268.45it/s]


Now have this 2x2 numpy array... turn it into a pandas DataFrame using the column names we saved earlier

In [8]:
mergedDF=pd.DataFrame(combined,columns=allCols)
mergedDF

Shelfmark Parent Label System number Volume enumeration  \
0     Arundel MS 54                                                  
1     Arundel MS 22                                                  
2      Arundel MS 5                                                  
3       Arundel MS 8                                                 
4      Arundel MS 17                                                 
...              ...    ...   ...           ...                ...   
7629     MSS Jav 24                                                  
7630  Sloane MS 2645                                                 
7631        Or 16875                                                 
7632    Add MS 12398                                                 
7633    Mss Eur D742                                                 

     Project (if part of a Project) DAR number  \
0                                                
1                                                
2                                                
3                                                
4                                                
...                             ...        ...   
7629  Henry Ginsburg Legacy Phase 2    DAR 501   
7630  Henry Ginsburg Legacy Phase 2    DAR 501   
7631                                             
7632                                             
7633                                             

                                          Source folder Output Folder  \
0     \\ad\collections\AAS Storage\Avestan (USW)\Aru...                 
1     \\ad\collections\AAS Storage\Avestan (USW)\Ave...                 
2     \\ad\collections\AAS Storage\Avestan (USW)\Ave...                 
3     \\ad\collections\AAS Storage\Avestan (USW)\ave...                 
4     \\ad\collections\AAS Storage\Avestan (USW)\Mss...                 
...                                                 ...           ...   
7629  \\ad\collections\Singapore-Malay\Ginsburg 2014...                 
7630  \\ad\collections\Singapore-Malay\Ginsburg 2014...                 
7631  \\ad\collections\Singapore-Malay\or 16875-unco...                 
7632  \\ad\collections\Singapore-Malay\SJB documents...                 
7633  \\ad\collections\Singapore-Malay\SJB documents...                 

     Ingest Format  ... IAMS - Institution identifier rules  \
0                   ...               ARCHON (Archive Desc)   
1                   ...               ARCHON (Archive Desc)   
2                   ...               ARCHON (Archive Desc)   
3                   ...               ARCHON (Archive Desc)   
4                   ...               ARCHON (Archive Desc)   
...            ...  ...                                 ...   
7629                ...               ARCHON (Archive Desc)   
7630                ...               ARCHON (Archive Desc)   
7631                ...               ARCHON (Archive Desc)   
7632                ...               ARCHON (Archive Desc)   
7633                ...               ARCHON (Archive Desc)   

       IAMS - Countries rules   IAMS - Structure rules  \
0     ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
1     ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
2     ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
3     ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
4     ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
...                       ...                      ...   
7629  ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
7630  ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
7631  ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
7632  ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   
7633  ISO 3166 (Archive Desc)  ISAD(G)2 (Archive Desc)   

           IAMS - Dates rules IAMS - Name rules  \
0     ISO 8601 (Archive Desc)                     
1     ISO 8601 (Archive Desc)                     
2     ISO 8601 (Archive Desc)                     
3     ISO 8601 (Archive Desc)     

Then, optionally, write this to a file:

In [9]:
mergedDF.to_excel("AAC_CombinedAnalysis_MergedIAMS_jN.xlsx",index=False)

N.B. this will produce an excel file with absolutely no formatting, which is probably not wanted. For now I've just used this as a way to make an excel file with the exact same rows as the original AAC_CombinedAnalysis spreadsheet so that the extra IAMS information can be copy-pasted directly into the original spreadsheet, which retains the formatting.

### To-do in this section: 
Task - Priority
- Adjust absolute folder paths for those that were migrated to new servers - medium
- Combine Aleph metadata into Master Spreadsheet (after Jess confirms shelfmarks/system numbers) - low 
- Combine Customer Order information into Master Spreadsheet - low
- Concatenate the two Notes columns, separated by ; - low


- Combine Siegfried output into master spreadsheet - are we still trying this, or keeping Siegfried separate?

## Corruption & data loss

### Which files have file size of 0 MB?

Only three files have a size less than 1 KB, and none have a size of 0 MB.

In [13]:
allData[allData.SIZE<1000] # size in bytes

ID PARENT_ID                                                URI  \
595161  646078            file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   
595223  646140            file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   
595312  646229            file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   

                                                FILE_PATH       NAME  \
595161  \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...  00004.tif   
595223  \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...  00066.tif   
595312  \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...  00155.tif   

           METHOD STATUS   SIZE  TYPE  EXT         LAST_MODIFIED  \
595161  Signature   Done  668.0  File  tif  2016-03-09T11:49:29Z   
595223  Signature   Done  626.0  File  tif  2016-03-09T11:51:39Z   
595312  Signature   Done  642.0  File  tif  2016-03-09T11:52:43Z   

        EXTENSION_MISMATCH                                        SHA256_HASH  \
595161               False  3ec734de8906f949d2092c76a1da28cc9c84ed221049a4...   
595223               False  437f28c852b1e1d1b6937b87014767f0b5398c5c4e44ec...   
595312               False  14345af01ba3eeeae71eeb6e6b45d9cbd96dba8124614c...   

        FORMAT_COUNT     PUID   MIME_TYPE               FORMAT_NAME  \
595161           1.0  fmt/353  image/tiff  Tagged Image File Format   
595223           1.0  fmt/353  image/tiff  Tagged Image File Format   
595312           1.0  fmt/353  image/tiff  Tagged Image File Format   

       FORMAT_VERSION                                         PARENT_DIR  \
595161                 \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
595223                 \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
595312                 \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   

        FolderCount  
595161            1  
595223            1  
595312            1

### Which files have a file mismatch? File extension = true

Only one file identified as tif by MIME_TYPE is affected

In [12]:
allData[allData['EXTENSION_MISMATCH']]

ID PARENT_ID                                                URI  \
740608  791881            file:///P8L-NAS2/Singapore-Malay/Finished%20Ma...   

                                                FILE_PATH         NAME  \
740608  \\P8L-NAS2\Singapore-Malay\Finished Manuscript...  ~psB4B4.tmp   

           METHOD STATUS         SIZE  TYPE  EXT         LAST_MODIFIED  \
740608  Signature   Done  102065984.0  File  tmp  2015-03-12T17:23:02Z   

        EXTENSION_MISMATCH                                        SHA256_HASH  \
740608                True  ff602f7eeb24442c251f7f573e9465a9f8bb7277b2f577...   

        FORMAT_COUNT       PUID   MIME_TYPE  \
740608           1.0  x-fmt/387  image/tiff   

                                          FORMAT_NAME FORMAT_VERSION  \
740608  Exchangeable Image File Format (Uncompressed)            2.2   

                                               PARENT_DIR  FolderCount  
740608  \\P8L-NAS2\Singapore-Malay\Finished Manuscript...            1

### For items: 
### - In the folders containing exact duplicates, which of these folders also contain not-exact duplicates? - medium
### - In the folders containing exact duplicates, which of these folders have a mismatch re number of files? - medium

Using the entire Siegfried output, find all duplicates (including the original) with the same method used above. After this, group the output by parent folder with `groupby`. This produces one row per parent folder, so all other columns have some operation applied to them. In this case they are aggregated - simply summed if the data type is `float64` or concatenated into a single string in any other case. For `SHA256_HASH`, this will produce a single hash from all items in a folder. If two parent directories have the same combined hash, they are full duplicates of each other, if not there is some difference with one or more files within a folder.

In [27]:
dao=allData[allData.duplicated(subset="SHA256_HASH",keep=False)]
DupFolders=dao.groupby(['PARENT_DIR'],as_index=False).agg(lambda x : x.sum() if x.dtype=='float64' else ' '.join(x))
DupFolders

PARENT_DIR  \
0     \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1     \\P12L-NAS5\Persian Digitisation\IO Islamic 10...   
2     \\P12L-NAS5\Persian Digitisation\IO Islamic 12...   
3     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...   
4     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...   
...                                                 ...   
1286  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1287  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1288  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1289  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1290  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                                              PARENT_ID  \
0                                                   ...   
1                                                   ...   
2                                                   ...   
3                                                   ...   
4                                                   ...   
...                                                 ...   
1286                                                ...   
1287                                                      
1288                                                      
1289                                                      
1290                                                ...   

                                                    URI  \
0     file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1     file:///P12L-NAS5/Persian%20Digitisation/IO%20...   
2     file:///P12L-NAS5/Persian%20Digitisation/IO%20...   
3     file:///P12L-NAS5/Persian%20Digitisation/IO%20...   
4     file:///P12L-NAS5/Persian%20Digitisation/IO%20...   
...                                                 ...   
1286  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1287  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1288  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1289  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1290  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                              FILE_PATH  \
0     \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1     \\P12L-NAS5\Persian Digitisation\IO Islamic 10...   
2     \\P12L-NAS5\Persian Digitisation\IO Islamic 12...   
3     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...   
4     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...   
...                                                 ...   
1286  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1287  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1288  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1289  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1290  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                                                   NAME  \
0     grenville_xli_f001r.tif grenville_xli_f001v.ti...   
1     io_islamic_1026_f001r.tif io_islamic_1026_f001...   
2     i_o_islamic_1256_control1.tif i_o_islamic_1256...   
3     io_islamic_132_f001r.tif io_islamic_132_f001v....   
4     io_islamic_137_f001r.tif io_islamic_137_f001v....   
...                                                 ...   
1286  or_16790_f001r.tif or_16790_f001v.tif or_16790...   
1287  or_16893_f001r.tif or_16893_f001v.tif or_16893...   
1288  OR_4736_f001r.tif OR_4736_f001v.tif OR_4736_f0...   
1289  or_4762_f001r.tif or_4762_f003r.tif or_4762_f0...   
1290  or_7458_f001r.tif or_7458_f001v.tif or_7458_f0...   

                                                 METHOD  \
0     Signature Signature Signature Signature Signat...   
1     Signature Signature Signature Signature Signat...   
2     Signature Signature Signature Signature Signat...   
3     Signature Signature Signature Signature Signat...   
4     Signature Signature Signature Signature Signat...   
...                                                 ...   
1286  Signature Signature Signature Signature Signat...   
1287  Signature Signature Si

In [34]:
DupFolders[DupFolders.duplicated(subset="SHA256_HASH",keep=False)].sort_values(by='SHA256_HASH')

PARENT_DIR  \
1043  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
496   \\P8L-NAS2\AAS Storage\Syriac (IT)\add_ms_1465...   
1261  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
406   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Th...   
97    \\P8L-NAS2\AAS Storage\AAS Legacy Images\io_sa...   
...                                                 ...   
1169  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
647   \\P8L-NAS3\Legacy Digital Content\Portfolio - ...   
590   \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
58     \\P12L-NAS5\Persian Digitisation\or 4615\or_4615   
1118  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   

                                              PARENT_ID  \
1043                                                ...   
496                                                 ...   
1261                                                      
406                                                       
97                                                  ...   
...                                                 ...   
1169                                                      
647                                                 ...   
590                                                 ...   
58                                                  ...   
1118                                                ...   

                                                    URI  \
1043  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
496   file:///P8L-NAS2/AAS%20Storage/Syriac%20%28IT%...   
1261  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
406   file:///P8L-NAS2/AAS%20Storage/Southeast%20Asi...   
97    file:///P8L-NAS2/AAS%20Storage/AAS%20Legacy%20...   
...                                                 ...   
1169  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
647   file:///P8L-NAS3/Legacy%20Digital%20Content/Po...   
590   file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   
58    file:///P12L-NAS5/Persian%20Digitisation/or%20...   
1118  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                              FILE_PATH  \
1043  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
496   \\P8L-NAS2\AAS Storage\Syriac (IT)\add_ms_1465...   
1261  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
406   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Th...   
97    \\P8L-NAS2\AAS Storage\AAS Legacy Images\io_sa...   
...                                                 ...   
1169  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
647   \\P8L-NAS3\Legacy Digital Content\Portfolio - ...   
590   \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
58    \\P12L-NAS5\Persian Digitisation\or 4615\or_46...   
1118  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   

                                                   NAME  \
1043  add_ms_14654_f001r.tif add_ms_14654_f001v.tif ...   
496   add_ms_14654_f001r.tif add_ms_14654_f001v.tif ...   
1261  or.15749!16.2_f001r.tif or.15749!16.2_f001v.ti...   
406   or.15749!16.2_f001r.tif or.15749!16.2_f001v.ti...   
97    io_san_3177_f001r.tif io_san_3177_f001v.tif io...   
...                                                 ...   
1169  or.15749!12.13_f001r.tif or.15749!12.13_f001v....   
647   003909.tif 011792.tif 011793.TIF 011796.tif 01...   
590   003909.tif 011792.tif 011793.TIF 011796.tif 01...   
58    or_4615_f001r.tif or_4615_f001v.tif or_4615_f0...   
1118  or_4615_f001r.tif or_4615_f001v.tif or_4615_f0...   

                                                 METHOD  \
1043  Signature Signature Signature Signature Signat...   
496   Signature Signature Signature Signature Signat...   
1261                      Signature Signature Signature   
406                       Signature Signature Signature   
97    Signature Signature Signature Signature Signat...   
...                                                 ...   
1169  Signature Signature Signature Signature Signat...   
647   Signature Signature Si

Check a couple of partial hashes to see if an exact duplicate is found

In [35]:
DupFolders[DupFolders['SHA256_HASH'].str.contains("ffd0ff02de9da11e46ecd4209e3eee214ee3baa62f488")]

PARENT_DIR PARENT_ID  \
446   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Th...             
1169  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...             

                                                    URI  \
446   file:///P8L-NAS2/AAS%20Storage/Southeast%20Asi...   
1169  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                              FILE_PATH  \
446   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Th...   
1169  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                                                   NAME  \
446   or.15749!12.13_f001r.tif or.15749!12.13_f001v....   
1169  or.15749!12.13_f001r.tif or.15749!12.13_f001v....   

                                                 METHOD  \
446   Signature Signature Signature Signature Signat...   
1169  Signature Signature Signature Signature Signat...   

                                  STATUS         SIZE  \
446   Done Done Done Done Done Done Done  657882208.0   
1169  Done Done Done Done Done Done Done  657882208.0   

                                    TYPE                          EXT  \
446   File File File File File File File  tif tif tif tif tif tif tif   
1169  File File File File File File File  tif tif tif tif tif tif tif   

                                          LAST_MODIFIED  \
446   2010-04-15T15:29:53+01:00 2010-04-15T15:30:05+...   
1169  2010-04-15T15:29:53+01:00 2010-04-15T15:30:05+...   

                                            SHA256_HASH  FORMAT_COUNT  \
446   ffd0ff02de9da11e46ecd4209e3eee214ee3baa62f4889...           7.0   
1169  ffd0ff02de9da11e46ecd4209e3eee214ee3baa62f4889...           7.0   

                                                   PUID  \
446   x-fmt/387 x-fmt/387 x-fmt/387 x-fmt/387 x-fmt/...   
1169  x-fmt/387 x-fmt/387 x-fmt/387 x-fmt/387 x-fmt/...   

                                              MIME_TYPE  \
446   image/tiff image/tiff image/tiff image/tiff im...   
1169  image/tiff image/tiff image/tiff image/tiff im...   

                                            FORMAT_NAME  
446   Exchangeable Image File Format (Uncompressed) ...  
1169  Exchangeable Image File Format (Uncompressed) ...

Happy that this methodology is sort of working, we can go on and list the folders that do not have an exact duplicate.

In [36]:
DupFolders[~DupFolders.duplicated(subset="SHA256_HASH",keep=False)].sort_values(by='SHA256_HASH')

PARENT_DIR  \
497   \\P8L-NAS2\AAS Storage\Syriac (IT)\add_ms_1465...   
1011  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
835   \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
716   \\P8l-NAS2\Photographic Studio data\Sent to Bo...   
104   \\P8L-NAS2\AAS Storage\AAS Legacy Images\or_12174   
...                                                 ...   
279   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Or ...   
736   \\P8l-NAS2\Photographic Studio data\Sent to Bo...   
230   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
346   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Gi...   
1278  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                                              PARENT_ID  \
497                                                 ...   
1011                                                      
835                                                 ...   
716                                                       
104                                                       
...                                                 ...   
279                                                       
736                                                       
230                                                       
346                                                 ...   
1278                                                ...   

                                                    URI  \
497   file:///P8L-NAS2/AAS%20Storage/Syriac%20%28IT%...   
1011  file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
835   file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
716   file:///P8l-NAS2/Photographic%20Studio%20data/...   
104   file:///P8L-NAS2/AAS%20Storage/AAS%20Legacy%20...   
...                                                 ...   
279   file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
736   file:///P8l-NAS2/Photographic%20Studio%20data/...   
230   file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
346   file:///P8L-NAS2/AAS%20Storage/Southeast%20Asi...   
1278  file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                              FILE_PATH  \
497   \\P8L-NAS2\AAS Storage\Syriac (IT)\add_ms_1465...   
1011  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
835   \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
716   \\P8l-NAS2\Photographic Studio data\Sent to Bo...   
104   \\P8L-NAS2\AAS Storage\AAS Legacy Images\or_12...   
...                                                 ...   
279   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Or ...   
736   \\P8l-NAS2\Photographic Studio data\Sent to Bo...   
230   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
346   \\P8L-NAS2\AAS Storage\Southeast Asia (ATG)\Gi...   
1278  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                                                   NAME  \
497   add_ms_14654_f01r.tif add_ms_14654_f01v.tif ad...   
1011  or_4726_fblefr.tif or_4726_fbrigv.tif or_4726_...   
835   or_2884_f001r.tif or_2884_f001v.tif or_2884_f0...   
716                         bl06_1_00.tif bl06_1_01.tif   
104   or_12174_f001r.tif or_12174_f002r.tif or_12174...   
...                                                 ...   
279   Or.4769, f.1.tif Or.4769, f.10.tif Or.4769, f....   
736   bl12_2_071.tif bl12_2_072.tif bl12_2_073.tif b...   
230   Dardic_1.tif Dardic_2.tif Dardic_3.tif Dardic_...   
346   Or_14844_f01r.tif Or_14844_f01v.tif Or_14844_f...   
1278  or_14844_f001r.tif or_14844_f001v.tif or_14844...   

                                                 METHOD  \
497   Signature Signature Signature Signature Signat...   
1011                      Signature Signature Signature   
835   Signature Signature Signature Signature Signat...   
716                                 Signature Signature   
104   Signature Signature Signature Signature Signat...   
...                                                 ...   
279   Signature Signature Signature Signature Signat...   
736   Signature Signature Si

This shows 389 folders that do not contain exact duplicates (or potentially have a different number of files within the folder)

### - In the folders containing sets of images, are any files significantly smaller than the others? (threshold?) - medium 

The approach I took for this was to group the original data by parent folder, find the number of files in each folder `nFilesInFolder` and then assign this to to the original data, although not writing anything out to a file

In [77]:
allDataCopy=allData
nFiles=allDataCopy.groupby(by="PARENT_DIR").sum()
allDataCopy=allDataCopy.join(allDataCopy.groupby('PARENT_DIR')['FolderCount'].sum(), on='PARENT_DIR',rsuffix="_FULL")
allDataCopy=allDataCopy.join(allDataCopy.groupby('PARENT_DIR')['SIZE'].sum(), on='PARENT_DIR',rsuffix="_TOTAL")
allDataCopy=allDataCopy.join(allDataCopy.groupby('PARENT_DIR')['SIZE'].mean(), on='PARENT_DIR',rsuffix="_MEAN")
allDataCopy=allDataCopy.join(allDataCopy.groupby('PARENT_DIR')['SIZE'].median(), on='PARENT_DIR',rsuffix="_MEDIAN")
allDataCopy

ID PARENT_ID                                                URI  \
0              1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1              2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2              3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3              4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4              5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
...          ...       ...                                                ...   
1276169  1328567            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1276170  1328568            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1276171  1328569            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1276172  1328570            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1276173  1328571            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
0        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
2        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
3        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
4        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
...                                                    ...   
1276169  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1276170  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1276171  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1276172  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1276173  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                            NAME     METHOD STATUS         SIZE  TYPE  EXT  \
0        grenville_xli_f001r.tif  Signature   Done   90714782.0  File  tif   
1        grenville_xli_f001v.tif  Signature   Done   85295734.0  File  tif   
2        grenville_xli_f002r.tif  Signature   Done   89390472.0  File  tif   
3        grenville_xli_f002v.tif  Signature   Done   82452820.0  File  tif   
4        grenville_xli_f003r.tif  Signature   Done   86180984.0  File  tif   
...                          ...        ...    ...          ...   ...  ...   
1276169        OR_8851_f067v.tif  Signature   Done   93801282.0  File  tif   
1276170        OR_8851_f068r.tif  Signature   Done   93801282.0  File  tif   
1276171       OR_8851_fblefr.tif  Signature   Done   87091258.0  File  tif   
1276172       OR_8851_fblefv.tif  Signature   Done   86587348.0  File  tif   
1276173        or_9770_f001r.tif  Signature   Done  118331318.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
0        2013-04-22T10:40:14+01:00               False   
1        2013-04-25T10:57:52+01:00               False   
2        2013-04-22T10:40:32+01:00               False   
3        2013-04-22T11:11:26+01:00               False   
4        2013-04-22T10:40:44+01:00               False   
...                            ...                 ...   
1276169       2011-11-16T11:57:02Z               False   
1276170       2011-11-16T11:57:19Z               False   
1276171       2011-11-16T10:55:54Z               False   
1276172       2011-11-16T11:59:16Z               False   
1276173       2013-02-18T11:44:07Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
0        8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...           1.0   
1        256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...           1.0   
2        9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...           1.0   
3        a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...           1.0   
4        9429574137662b13efa5256290a75982f9c3ec60245233...           1.0   
...                                                    ...           ...   
1276169  85336472039fa38ec4a7c83eaebab4f78187029f84c65f...           1.0   
1276170  597985b2962bf72daba690f86ecc4a3f2f649fa4f0fa7

Print examples of cases where a file in a directory is less than 10% of the mean file size:

In [80]:
allDataCopy[allDataCopy['SIZE']<=0.01*allDataCopy['SIZE_MEAN']]

ID PARENT_ID                                                URI  \
175406    179188            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
201713    206347            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201729    206363            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201734    206368            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201748    206382            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
...          ...       ...                                                ...   
1252505  1304726            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252519  1304740            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252528  1304749            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252555  1304776            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252564  1304785            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
175406   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
201713   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201729   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201734   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201748   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
...                                                    ...   
1252505  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252519  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252528  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252555  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252564  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   

                             NAME     METHOD STATUS     SIZE  TYPE  EXT  \
175406      ORC_BEN_PBS-5_001.tif  Signature   Done   6765.0  File  tif   
201713    mss_eur_e207!5_f004.TIF  Signature   Done  26146.0  File  TIF   
201729    mss_eur_e207!5_f020.TIF  Signature   Done  23588.0  File  TIF   
201734    mss_eur_e207!5_f025.TIF  Signature   Done  21024.0  File  TIF   
201748    mss_eur_e207!5_f039.TIF  Signature   Done  20776.0  File  TIF   
...                           ...        ...    ...      ...   ...  ...   
1252505  mss_eur_e207!5_f104r.TIF  Signature   Done  16168.0  File  TIF   
1252519  mss_eur_e207!5_f118r.TIF  Signature   Done   8098.0  File  TIF   
1252528  mss_eur_e207!5_f127r.TIF  Signature   Done  16936.0  File  TIF   
1252555  mss_eur_e207!5_f154r.TIF  Signature   Done  20988.0  File  TIF   
1252564  mss_eur_e207!5_f163r.TIF  Signature   Done  19164.0  File  TIF   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
175406   2015-08-07T09:50:08+01:00               False   
201713        2013-11-13T08:27:00Z               False   
201729        2013-11-13T08:29:50Z               False   
201734        2013-11-13T08:35:56Z               False   
201748        2013-11-13T08:38:49Z               False   
...                            ...                 ...   
1252505       2013-11-13T09:01:19Z               False   
1252519       2013-11-13T09:04:46Z               False   
1252528       2013-11-13T09:06:34Z               False   
1252555       2013-11-13T09:09:52Z               False   
1252564       2013-11-13T09:27:41Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
175406   3f9f3750e017ccf02fb3b8cb501b7445bb0a05be0f1456...           1.0   
201713   97ec827a8d577a7157f164d8106b482bef363e037b1a05...           1.0   
201729   77876580f0ec5bdc05ece1d192a0881225036f8b499b5f...           1.0   
201734   7bde3801ce88f3897c4a5dbe6cd3b7b8a1cc7b1e53d398...           1.0   
201748   6baf6ec458e9b64e6d9835aa6a280336e95128146ebc38...           1.0   
...                                                    ...           ...   
1252505  c0cee68a85bb8777ffc0055c6729dcfc6cc5c05dd8c3fe...           1.0   
1252519  621e6d9c8061f7a8b23b4b679fee9ae9de7eaaabe8b398...           1.0   
1252528  9f9be

In [79]:
allDataCopy[allDataCopy['SIZE']<=0.01*allDataCopy['SIZE_MEDIAN']]

ID PARENT_ID                                                URI  \
175406    179188            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
201713    206347            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201729    206363            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201734    206368            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
201748    206382            file:///P8L-NAS2/AAS%20Storage/Persian%20%28US...   
...          ...       ...                                                ...   
1252505  1304726            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252519  1304740            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252528  1304749            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252555  1304776            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1252564  1304785            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
175406   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
201713   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201729   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201734   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
201748   \\P8L-NAS2\AAS Storage\Persian (USW - MIW)\Mss...   
...                                                    ...   
1252505  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252519  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252528  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252555  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   
1252564  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Persian...   

                             NAME     METHOD STATUS     SIZE  TYPE  EXT  \
175406      ORC_BEN_PBS-5_001.tif  Signature   Done   6765.0  File  tif   
201713    mss_eur_e207!5_f004.TIF  Signature   Done  26146.0  File  TIF   
201729    mss_eur_e207!5_f020.TIF  Signature   Done  23588.0  File  TIF   
201734    mss_eur_e207!5_f025.TIF  Signature   Done  21024.0  File  TIF   
201748    mss_eur_e207!5_f039.TIF  Signature   Done  20776.0  File  TIF   
...                           ...        ...    ...      ...   ...  ...   
1252505  mss_eur_e207!5_f104r.TIF  Signature   Done  16168.0  File  TIF   
1252519  mss_eur_e207!5_f118r.TIF  Signature   Done   8098.0  File  TIF   
1252528  mss_eur_e207!5_f127r.TIF  Signature   Done  16936.0  File  TIF   
1252555  mss_eur_e207!5_f154r.TIF  Signature   Done  20988.0  File  TIF   
1252564  mss_eur_e207!5_f163r.TIF  Signature   Done  19164.0  File  TIF   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
175406   2015-08-07T09:50:08+01:00               False   
201713        2013-11-13T08:27:00Z               False   
201729        2013-11-13T08:29:50Z               False   
201734        2013-11-13T08:35:56Z               False   
201748        2013-11-13T08:38:49Z               False   
...                            ...                 ...   
1252505       2013-11-13T09:01:19Z               False   
1252519       2013-11-13T09:04:46Z               False   
1252528       2013-11-13T09:06:34Z               False   
1252555       2013-11-13T09:09:52Z               False   
1252564       2013-11-13T09:27:41Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
175406   3f9f3750e017ccf02fb3b8cb501b7445bb0a05be0f1456...           1.0   
201713   97ec827a8d577a7157f164d8106b482bef363e037b1a05...           1.0   
201729   77876580f0ec5bdc05ece1d192a0881225036f8b499b5f...           1.0   
201734   7bde3801ce88f3897c4a5dbe6cd3b7b8a1cc7b1e53d398...           1.0   
201748   6baf6ec458e9b64e6d9835aa6a280336e95128146ebc38...           1.0   
...                                                    ...           ...   
1252505  c0cee68a85bb8777ffc0055c6729dcfc6cc5c05dd8c3fe...           1.0   
1252519  621e6d9c8061f7a8b23b4b679fee9ae9de7eaaabe8b398...           1.0   
1252528  9f9be

### In the folders containing duplicate sets of images, do any of the folders contain a different number of TIFF files?

This assigns a unique match ID to files with matching SHA256 hashes. Can loop through each match ID and find the number of files in each of the folders.

In [115]:
#dao.groupby(['SHA256_HASH'],as_index=False).agg(lambda x : x.sum() if x.dtype=='float64' else ';'.join(x))
groupedHash=dao.groupby('SHA256_HASH').sum()
groupedHash['SHA256_HASH']=groupedHash.index

groupedHash.index=np.arange(0,len(groupedHash))


uniqueID=[]
for index, row in groupedHash.iterrows():
    matchID="matchID_"+str(index)
    uniqueID.append(matchID)

groupedHash['MatchID']=uniqueID
hashMatch=groupedHash[['SHA256_HASH','MatchID']]

In [117]:
allDataCopy=allDataCopy.merge(hashMatch, on='SHA256_HASH', how='left') 

Now merge the MatchID into the SHA256 duplicates dataframe

In [140]:
dao=allDataCopy[allDataCopy.duplicated(subset="SHA256_HASH",keep=False)]
dao=dao.sort_values(by="MatchID")
dao

ID PARENT_ID                                                URI  \
0              1            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
1              2            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
2              3            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
3              4            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
4              5            file:///P12L-NAS5/Persian%20Digitisation/Grenv...   
...          ...       ...                                                ...   
1275916  1328311            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275917  1328312            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275918  1328313            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275919  1328314            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
1275920  1328315            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   

                                                 FILE_PATH  \
0        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
1        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
2        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
3        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
4        \\P12L-NAS5\Persian Digitisation\Grenville xli...   
...                                                    ...   
1275916  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275917  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275918  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275919  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
1275920  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   

                            NAME     METHOD STATUS        SIZE  TYPE  EXT  \
0        grenville_xli_f001r.tif  Signature   Done  90714782.0  File  tif   
1        grenville_xli_f001v.tif  Signature   Done  85295734.0  File  tif   
2        grenville_xli_f002r.tif  Signature   Done  89390472.0  File  tif   
3        grenville_xli_f002v.tif  Signature   Done  82452820.0  File  tif   
4        grenville_xli_f003r.tif  Signature   Done  86180984.0  File  tif   
...                          ...        ...    ...         ...   ...  ...   
1275916       or_7458_fbrigr.tif  Signature   Done  66538396.0  File  tif   
1275917       or_7458_fbrigv.tif  Signature   Done  66538396.0  File  tif   
1275918        or_7458_fbspi.tif  Signature   Done  39485948.0  File  tif   
1275919      or_7458_fse002r.tif  Signature   Done  66538396.0  File  tif   
1275920      or_7458_fse002v.tif  Signature   Done  66538396.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
0        2013-04-22T10:40:14+01:00               False   
1        2013-04-25T10:57:52+01:00               False   
2        2013-04-22T10:40:32+01:00               False   
3        2013-04-22T11:11:26+01:00               False   
4        2013-04-22T10:40:44+01:00               False   
...                            ...                 ...   
1275916       2007-01-04T00:42:21Z               False   
1275917       2007-01-04T00:44:31Z               False   
1275918       2007-01-04T00:46:28Z               False   
1275919       2007-01-04T00:38:13Z               False   
1275920       2007-01-04T00:40:08Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
0        8a0a01aa68bdf8e39604b747829f4c7fa38755ce335c3f...           1.0   
1        256f682db10c7b57f5aaf4257bfa8220388383e8d835d2...           1.0   
2        9320ca9074e19ad5537d1dc4b78d3574916c4d56bcc5f4...           1.0   
3        a71db4bbd406b34f7e407240030a5eb82d72955ecc6e27...           1.0   
4        9429574137662b13efa5256290a75982f9c3ec60245233...           1.0   
...                                                    ...           ...   
1275916  0c9f2baf3dd8477bd20327df6f0d3733c8392878512110...           1.0   
1275917  8cc44b99495ca7e7767600d1ba26edb385fa8232826978...        

This groups the dataframe by the MatchID and shows the absolute difference (always displayed as a positive number) between the number of files in the folders containing files with the same SHA256 hash

In [146]:
dao['diff'] = dao.groupby('MatchID')['FolderCount_FULL'].diff().abs()

Now show cases where the number of files is different

In [157]:
dao[dao['diff']>0.0]

ID PARENT_ID                                                URI  \
178267    182050            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
902908    954973            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
456980    507492            file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   
87235      87674            file:///P8L-NAS2/AAS%20Storage/AAS%20Legacy%20...   
177440    181222            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
...          ...       ...                                                ...   
180904    184711            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
78296      78733            file:///P8L-NAS2/AAS%20Storage/AAS%20Legacy%20...   
1275779  1328174            file:///v19l-nas1/Greek%20MSS%20Phase%202%20Fo...   
162652    166434            file:///P8L-NAS2/AAS%20Storage/Korean%20%28HT%...   
178396    182179            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   

                                                 FILE_PATH  \
178267   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
902908   \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
456980   \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
87235    \\P8L-NAS2\AAS Storage\AAS Legacy Images\or_12...   
177440   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
...                                                    ...   
180904   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
78296    \\P8L-NAS2\AAS Storage\AAS Legacy Images\add_m...   
1275779  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   
162652   \\P8L-NAS2\AAS Storage\Korean (HT)\Or 7458\TIF...   
178396   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   

                           NAME     METHOD STATUS        SIZE  TYPE  EXT  \
178267               000811.tif  Signature   Done    146948.0  File  tif   
902908       or_ 2705_f212r.tif  Signature   Done  91971630.0  File  tif   
456980           bl12_2_062.tif  Signature   Done  42364884.0  File  tif   
87235        or_12439_f080v.tif  Signature   Done  42364884.0  File  tif   
177440               001228.tif  Signature   Done     47988.0  File  tif   
...                         ...        ...    ...         ...   ...  ...   
180904           Bihari_614.tif  Signature   Done    148256.0  File  tif   
78296    add_ms_11558_f106v.tif  Signature   Done  85973024.0  File  tif   
1275779       or_7458_f027r.tif  Signature   Done  66538396.0  File  tif   
162652        or_7458_f027r.tif  Signature   Done  66538396.0  File  tif   
178396               000940.tif  Signature   Done    148392.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
178267   2015-08-07T13:50:22+01:00               False   
902908   2014-07-18T10:36:34+01:00               False   
456980   2002-10-01T10:19:13+01:00               False   
87235    2002-10-01T10:19:13+01:00               False   
177440   2015-08-07T12:12:14+01:00               False   
...                            ...                 ...   
180904   2015-08-07T12:11:39+01:00               False   
78296    2010-07-15T16:42:36+01:00               False   
1275779       2007-01-03T19:56:06Z               False   
162652        2007-01-03T19:56:06Z               False   
178396   2015-08-07T13:50:30+01:00               False   

                                               SHA256_HASH  FORMAT_COUNT  \
178267   00086e7b3bd406b5bada0c10a74a62d7b3a28c200df0ab...           1.0   
902908   0057424aa820f9f962f8deaec396c5ff64b2789ef5cfb3...           1.0   
456980   226a70af03b8785a17bfe0b55bc16be6b7e1dd4ebceeda...           1.0   
87235    226a70af03b8785a17bfe0b55bc16be6b7e1dd4ebceeda...           1.0   
177440   226d303b082df1de0c7f960a4a1c3e059bf2e83d2db278...           1.0   
...                                                    ...           ...   
180904   2246f147a5a7e5c3b169ab63fc0587e78c4b14d300fbd6...           1.0   
78296    22480f5e26e856b4efaf1c18bd2c2104eba2f5cd0148b6...           1.0   
12

Just as a reminder, this assigns each SHA256 value with at least one duplicate a unique matching ID. Therefore the folder containing this file can be considered to have been flagged as a duplicate folder - it may be flagged more than once by multiple matchID values. 

The 'diff' here compares the number of files in the parent folder of files with the same MatchID. This shows that for 10160 of these files the number of files contained in their parent folder is not consistent.

To see a more rounded picture, first get rid of any duplicate parent folders and sort by MatchID. Then look again at cases wher e `diff>0.0`

In [166]:
uniqueParentDirs=dao[~dao.duplicated(subset="PARENT_DIR")].sort_values(by="MatchID")
uniqueParentDirs[uniqueParentDirs['diff']>0.0]

ID PARENT_ID                                                URI  \
178267    182050            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
902908    954973            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
87235      87674            file:///P8L-NAS2/AAS%20Storage/AAS%20Legacy%20...   
456980    507492            file:///P8L-NAS2/AAS%20Storage/ZZZ%20-%20UNIDE...   
177440    181222            file:///P8L-NAS2/AAS%20Storage/Microfiches%20a...   
...          ...       ...                                                ...   
1172217  1224425            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
1170680  1222797            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
1151889  1203960            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
1172196  1224404            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   
1139945  1192016            file:///p12l-nas5/Hebrew%20MSS%20Digitisation/...   

                                                 FILE_PATH  \
178267   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
902908   \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
87235    \\P8L-NAS2\AAS Storage\AAS Legacy Images\or_12...   
456980   \\P8L-NAS2\AAS Storage\ZZZ - UNIDENTIFIED OR U...   
177440   \\P8L-NAS2\AAS Storage\Microfiches and card ca...   
...                                                    ...   
1172217  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
1170680  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
1151889  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
1172196  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   
1139945  \\p12l-nas5\Hebrew MSS Digitisation\Hebrew MSS...   

                               NAME     METHOD STATUS         SIZE  TYPE  EXT  \
178267                   000811.tif  Signature   Done     146948.0  File  tif   
902908           or_ 2705_f212r.tif  Signature   Done   91971630.0  File  tif   
87235            or_12439_f080v.tif  Signature   Done   42364884.0  File  tif   
456980               bl12_2_062.tif  Signature   Done   42364884.0  File  tif   
177440                   001228.tif  Signature   Done      47988.0  File  tif   
...                             ...        ...    ...          ...   ...  ...   
1172217        or_4800_stitched.tif  Signature   Done  488509984.0  File  tif   
1170680       or_11805_stitched.tif  Signature   Done  213425852.0  File  tif   
1151889          or_4798_fblefr.tif  Signature   Done   31824228.0  File  tif   
1172196  or_4799_f001r_STITCHED.tif  Signature   Done  574540428.0  File  tif   
1139945        or_1457_5_fblefr.tif  Signature   Done   31759636.0  File  tif   

                     LAST_MODIFIED  EXTENSION_MISMATCH  \
178267   2015-08-07T13:50:22+01:00               False   
902908   2014-07-18T10:36:34+01:00               False   
87235    2002-10-01T10:19:13+01:00               False   
456980   2002-10-01T10:19:13+01:00               False   
177440   2015-08-07T12:12:14+01:00               False   
...                            ...                 ...   
1172217  2015-08-17T08:57:01+01:00               False   
1170680  2015-10-13T10:20:54+01:00               False   
1151889       2016-02-15T16:42:27Z               False   
1172196  2015-08-21T14:06:45+01:00               False   
1139945       2016-02-05T13:16:15Z               False   

                                               SHA256_HASH  FORMAT_COUNT  \
178267   00086e7b3bd406b5bada0c10a74a62d7b3a28c200df0ab...           1.0   
902908   0057424aa820f9f962f8deaec396c5ff64b2789ef5cfb3...           1.0   
87235    226a70af03b8785a17bfe0b55bc16be6b7e1dd4ebceeda...           1.0   
456980   226a70af03b8785a17bfe0b55bc16be6b7e1dd4ebceeda...           1.0   
177440   226d303b082df1de0c7f960a4a1c3e059bf2e83d2db278...           1.0   
...                                                    ...           ...   
1172217  f0523d4d080b54da10b1136a4fcf98e8b167658d6a3e14...           1.0   
1170680  faf440627

Note that you still see some cases where the MatchID is the same, but `diff` varies. This indicates multiple duplicates of the same file, but answers the question set out above: 201 folders containing exact duplicates have a mismatch re: number of files.

N.B. the below is probably incorrect treatment and should be ignored for now:

In [191]:
#upd=uniqueParentDirs
#upd=upd.replace(np.nan,0.0)
#upd=upd[upd['diff']>0.]
#upd=upd.drop_duplicates(subset="MatchID")
#upd

~This now shows 187 folders, flagged as containing duplicates, that contain a different number of files than another folder flagged as containing one or more of the same duplicate files.~ 

^This was incorrect treatment of the problem and removed cases where an exact duplicate of a file was found in more than one location that contained a different number of files to other folders containing the same file. Reduced 201 cases to 187, when the upper estimate is more useful here

## DAMPS ingest

## To-do:
Task - priority (all low)
- ~Total number of tiff files in each folder?~
- ~Total tiff file size in each folder?~
- Identify all master and duplicate files(!) - in consultation with curators
- Identify all DAR projects and numbers
- Identify all sections responsible for folders
- Identify reading direction for all folders (map from language field?)

Already added a column of 1s to the original Siegfried output for each file, so grouping the data by parent directory and summing shows the number of tiff files in each folder in the `FolderCount` column. 

The size in bytes is shown for each folder in the `SIZE` column.

In [44]:
parentFolderAnalysis=allData.groupby(['PARENT_DIR'],as_index=False).sum()
parentFolderAnalysis

PARENT_DIR         ID  \
0     \\P12L-NAS5\Persian Digitisation\Grenville xli...     138075   
1     \\P12L-NAS5\Persian Digitisation\IO Islamic 10...     131670   
2     \\P12L-NAS5\Persian Digitisation\IO Islamic 12...    1733256   
3     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...     473424   
4     \\P12L-NAS5\Persian Digitisation\IO Islamic 13...    2645175   
...                                                 ...        ...   
7348  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...  257674389   
7349  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   71730495   
7350  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...   81032461   
7351  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...  182004774   
7352  \\v19l-nas1\Greek MSS Phase 2 Folder 2\Thai Ma...    1328571   

              SIZE  EXTENSION_MISMATCH  FORMAT_COUNT  FolderCount  
0     4.396242e+10               False         525.0          525  
1     3.393058e+10               False         209.0          209  
2     1.283186e+11               False        1267.0         1267  
3     2.354463e+10               False         224.0          224  
4     1.518890e+11               False         975.0          975  
...            ...                 ...           ...          ...  
7348  1.285657e+10               False         194.0          194  
7349  5.218129e+09               False          54.0           54  
7350  5.501632e+09               False          61.0           61  
7351  1.276355e+10               False         137.0          137  
7352  1.183313e+08               False           1.0            1  

[7353 rows x 6 columns]